File: internal/report/user_performance.go
package report

import (
	"fmt"
	"strings"
	"time"

	"github.com/tpdenta/afta-reception/internal/platform/apperror"
	"github.com/tpdenta/afta-reception/internal/user"
)

// userTypeSortOrder ترتیب نمایش انواع کاربر در گزارش (مطابق تعریف مدل).
var userTypeSortOrder = map[user.UserType]int{
	user.UserTypeUser:      0,
	user.UserTypeDoctor:    1,
	user.UserTypeAssistant: 2,
	user.UserTypeSpecialist: 3,
}

// parseUserPerformanceBase فیلترهای مشترک گزارش کارکرد کاربران را اعتبارسنجی و parse می‌کند.
func parseUserPerformanceBase(q UserPerformanceBaseQuery) (from, to time.Time, userIDs []uint, err error) {
	userIDs, err = parseUintList(q.UserIDs)
	if err != nil {
		return time.Time{}, time.Time{}, nil, apperror.New("VALIDATION_ERROR", "شناسه کاربر نامعتبر است.", "invalid user id", 400)
	}
	if len(userIDs) == 0 {
		return time.Time{}, time.Time{}, nil, apperror.New("VALIDATION_ERROR", "حداقل یک کاربر انتخاب کنید.", "user_ids required", 400)
	}

	from, err = combineDateTime(q.FromDate, q.FromTime, false)
	if err != nil {
		return time.Time{}, time.Time{}, nil, err
	}
	to, err = combineDateTime(q.ToDate, q.ToTime, true)
	if err != nil {
		return time.Time{}, time.Time{}, nil, err
	}
	if to.Before(from) {
		return time.Time{}, time.Time{}, nil, apperror.New("VALIDATION_ERROR", "تاریخ پایان نمی‌تواند قبل از تاریخ شروع باشد.", "invalid date range", 400)
	}
	return from, to, userIDs, nil
}

// buildUserPerformanceMeta متادیتای چاپ گزارش کارکرد کاربران را می‌سازد.
func (s *Service) buildUserPerformanceMeta(
	reportType string,
	fromDate, toDate, fromTime, toTime string,
	userIDs []uint,
	separateByPatient bool,
) (UserPerformanceMeta, error) {
	infos, err := s.repo.FindUserInfos(userIDs)
	if err != nil {
		return UserPerformanceMeta{}, apperror.New("DB_ERROR", "خطا در خواندن نام کاربران.", err.Error(), 500)
	}
	userNames := make([]string, 0, len(userIDs))
	for _, id := range userIDs {
		if info, ok := infos[id]; ok {
			userNames = append(userNames, strings.TrimSpace(info.Name+" "+info.Family))
		} else {
			userNames = append(userNames, fmt.Sprintf("کاربر %d", id))
		}
	}
	if fromTime == "" {
		fromTime = "00:00"
	}
	if toTime == "" {
		toTime = "23:59"
	}
	return UserPerformanceMeta{
		ReportType:        reportType,
		FromDate:          fromDate,
		ToDate:            toDate,
		FromTime:          fromTime,
		ToTime:            toTime,
		UserNames:         userNames,
		SeparateByPatient: separateByPatient,
	}, nil
}

// sortUserIDsByType کاربران را بر اساس نوع کاربر و سپس نام خانوادگی مرتب می‌کند.
func sortUserIDsByType(userIDs []uint, infos map[uint]userInfoRow) []uint {
	sorted := make([]uint, len(userIDs))
	copy(sorted, userIDs)
	type sorter struct {
		id   uint
		rank int
		name string
	}
	items := make([]sorter, 0, len(sorted))
	for _, id := range sorted {
		info, ok := infos[id]
		rank := 99
		name := fmt.Sprintf("کاربر %d", id)
		if ok {
			if r, exists := userTypeSortOrder[user.UserType(info.UserType)]; exists {
				rank = r
			}
			name = strings.TrimSpace(info.Name + " " + info.Family)
		}
		items = append(items, sorter{id: id, rank: rank, name: name})
	}
	for i := 0; i < len(items); i++ {
		for j := i + 1; j < len(items); j++ {
			swap := false
			if items[j].rank < items[i].rank {
				swap = true
			} else if items[j].rank == items[i].rank && items[j].name < items[i].name {
				swap = true
			}
			if swap {
				items[i], items[j] = items[j], items[i]
			}
		}
	}
	out := make([]uint, len(items))
	for i, item := range items {
		out[i] = item.id
	}
	return out
}

// userDisplayName نام نمایشی کاربر را برمی‌گرداند.
func userDisplayName(id uint, infos map[uint]userInfoRow) string {
	if info, ok := infos[id]; ok {
		return strings.TrimSpace(info.Name + " " + info.Family)
	}
	return fmt.Sprintf("کاربر %d", id)
}

// buildUserReceptionRow سطر تجمیعی پذیرش یک کاربر را می‌سازد.
func buildUserReceptionRow(userID uint, userName string, agg *userReceptionAggRow) UserPerformanceReceptionRow {
	row := UserPerformanceReceptionRow{
		UserID:             userID,
		UserName:           userName,
		ServiceAmount:      agg.ServiceAmount,
		ServiceTariff:      agg.ServiceTariff,
		OrganizationShare:  agg.OrganizationShare,
		SupplementaryShare: agg.SupplementaryShare,
		SubsidyShare:       agg.SubsidyShare,
		ReceptionCount:     agg.ReceptionCount,
	}
	row.PayableAmount = calcPayableAmount(row.ServiceAmount, row.OrganizationShare, row.SupplementaryShare, row.SubsidyShare)
	return row
}

// buildUserReceptionDetailRows سطرهای ریز پذیرش را می‌سازد.
func buildUserReceptionDetailRows(aggRows []userReceptionDetailAggRow) []UserPerformanceReceptionDetailRow {
	rows := make([]UserPerformanceReceptionDetailRow, 0, len(aggRows))
	for _, a := range aggRows {
		row := UserPerformanceReceptionDetailRow{
			ReceptionID:        a.ReceptionID,
			PatientName:        strings.TrimSpace(a.PatientFirstName + " " + a.PatientLastName),
			NationalCode:       a.NationalCode,
			ServiceAmount:      a.ServiceAmount,
			ServiceTariff:      a.ServiceTariff,
			OrganizationShare:  a.OrganizationShare,
			SupplementaryShare: a.SupplementaryShare,
			SubsidyShare:       a.SubsidyShare,
			ServicesReceived:   a.ServicesReceived,
		}
		row.PayableAmount = calcPayableAmount(row.ServiceAmount, row.OrganizationShare, row.SupplementaryShare, row.SubsidyShare)
		rows = append(rows, row)
	}
	return rows
}

// summarizeUserReceptionDetailRows جمع مالی ریز پذیرش‌ها را محاسبه می‌کند.
func summarizeUserReceptionDetailRows(rows []UserPerformanceReceptionDetailRow) UserPerformanceReceptionSummary {
	var summary UserPerformanceReceptionSummary
	summary.RowLabel = "جمع"
	for _, row := range rows {
		summary.ServiceAmount += row.ServiceAmount
		summary.ServiceTariff += row.ServiceTariff
		summary.OrganizationShare += row.OrganizationShare
		summary.SupplementaryShare += row.SupplementaryShare
		summary.SubsidyShare += row.SubsidyShare
		summary.PayableAmount += row.PayableAmount
	}
	summary.ReceptionCount = int64(len(rows))
	return summary
}

// mergeUserReceptionSummaries جمع چند خلاصه پذیرش را محاسبه می‌کند.
func mergeUserReceptionSummaries(summaries []UserPerformanceReceptionSummary) UserPerformanceReceptionSummary {
	out := UserPerformanceReceptionSummary{RowLabel: "جمع کل"}
	for _, s := range summaries {
		out.ServiceAmount += s.ServiceAmount
		out.ServiceTariff += s.ServiceTariff
		out.OrganizationShare += s.OrganizationShare
		out.SupplementaryShare += s.SupplementaryShare
		out.SubsidyShare += s.SubsidyShare
		out.PayableAmount += s.PayableAmount
		out.ReceptionCount += s.ReceptionCount
	}
	return out
}

// buildUserTransactionRow سطر تجمیعی تراکنش یک کاربر را می‌سازد.
func buildUserTransactionRow(userID uint, userName string, agg *userTransactionAggRow) UserPerformanceTransactionRow {
	return UserPerformanceTransactionRow{
		UserID:          userID,
		UserName:        userName,
		TotalReceived:   agg.TotalReceived,
		TotalPaid:       agg.TotalPaid,
		ReceivedTxCount: agg.ReceivedTxCount,
		PaidTxCount:     agg.PaidTxCount,
		TotalTxCount:    agg.TotalTxCount,
	}
}

// buildUserTransactionDetailRows سطرهای ریز تراکنش را می‌سازد.
func buildUserTransactionDetailRows(aggRows []userTransactionDetailAggRow) []UserPerformanceTransactionDetailRow {
	rows := make([]UserPerformanceTransactionDetailRow, 0, len(aggRows))
	for _, a := range aggRows {
		rows = append(rows, UserPerformanceTransactionDetailRow{
			TransactionID: a.TransactionID,
			FileNumber:    a.FileNumber,
			PatientName:   strings.TrimSpace(a.PatientFirstName + " " + a.PatientLastName),
			NationalCode:  a.NationalCode,
			Received:      a.Received,
			Paid:          a.Paid,
		})
	}
	return rows
}

// summarizeUserTransactionDetailRows جمع ریز تراکنش‌ها را محاسبه می‌کند.
func summarizeUserTransactionDetailRows(rows []UserPerformanceTransactionDetailRow) UserPerformanceTransactionSummary {
	var summary UserPerformanceTransactionSummary
	summary.RowLabel = "جمع"
	for _, row := range rows {
		summary.TotalReceived += row.Received
		summary.TotalPaid += row.Paid
		if row.Received > 0 {
			summary.ReceivedTxCount++
		}
		if row.Paid > 0 {
			summary.PaidTxCount++
		}
	}
	summary.TotalTxCount = int64(len(rows))
	return summary
}

// mergeUserTransactionSummaries جمع چند خلاصه تراکنش را محاسبه می‌کند.
func mergeUserTransactionSummaries(summaries []UserPerformanceTransactionSummary) UserPerformanceTransactionSummary {
	out := UserPerformanceTransactionSummary{RowLabel: "جمع کل"}
	for _, s := range summaries {
		out.TotalReceived += s.TotalReceived
		out.TotalPaid += s.TotalPaid
		out.ReceivedTxCount += s.ReceivedTxCount
		out.PaidTxCount += s.PaidTxCount
		out.TotalTxCount += s.TotalTxCount
	}
	return out
}

// UserPerformanceByReception گزارش کارکرد کاربران بر اساس پذیرش را برمی‌گرداند.
func (s *Service) UserPerformanceByReception(q UserPerformanceByReceptionQuery) (*UserPerformanceByReceptionResponse, error) {
	from, to, userIDs, err := parseUserPerformanceBase(q.UserPerformanceBaseQuery)
	if err != nil {
		return nil, err
	}

	infos, err := s.repo.FindUserInfos(userIDs)
	if err != nil {
		return nil, apperror.New("DB_ERROR", "خطا در خواندن اطلاعات کاربران.", err.Error(), 500)
	}
	sortedIDs := sortUserIDsByType(userIDs, infos)

	meta, err := s.buildUserPerformanceMeta("by_reception", q.FromDate, q.ToDate, q.FromTime, q.ToTime, sortedIDs, q.SeparateByPatient)
	if err != nil {
		return nil, err
	}

	if q.SeparateByPatient {
		sections := make([]UserPerformanceReceptionSection, 0, len(sortedIDs))
		sectionSummaries := make([]UserPerformanceReceptionSummary, 0, len(sortedIDs))
		for _, userID := range sortedIDs {
			info, ok := infos[userID]
			if !ok {
				continue
			}
			detailRows, err := s.repo.ListUserPerformanceReceptions(from, to, userID, user.UserType(info.UserType))
			if err != nil {
				return nil, apperror.New("DB_ERROR", "خطا در محاسبه گزارش.", err.Error(), 500)
			}
			receptionRows := buildUserReceptionDetailRows(detailRows)
			summary := summarizeUserReceptionDetailRows(receptionRows)
			sections = append(sections, UserPerformanceReceptionSection{
				UserID:   userID,
				UserName: userDisplayName(userID, infos),
				Rows:     receptionRows,
				Summary:  summary,
			})
			sectionSummaries = append(sectionSummaries, summary)
		}
		return &UserPerformanceByReceptionResponse{
			Summary:  mergeUserReceptionSummaries(sectionSummaries),
			Sections: sections,
			Meta:     meta,
		}, nil
	}

	rows := make([]UserPerformanceReceptionRow, 0, len(sortedIDs))
	var summary UserPerformanceReceptionSummary
	summary.RowLabel = "جمع کل"
	for _, userID := range sortedIDs {
		info, ok := infos[userID]
		if !ok {
			continue
		}
		agg, err := s.repo.AggregateUserPerformanceByReception(from, to, userID, user.UserType(info.UserType))
		if err != nil {
			return nil, apperror.New("DB_ERROR", "خطا در محاسبه گزارش.", err.Error(), 500)
		}
		row := buildUserReceptionRow(userID, userDisplayName(userID, infos), agg)
		rows = append(rows, row)
		summary.ServiceAmount += row.ServiceAmount
		summary.ServiceTariff += row.ServiceTariff
		summary.OrganizationShare += row.OrganizationShare
		summary.SupplementaryShare += row.SupplementaryShare
		summary.SubsidyShare += row.SubsidyShare
		summary.PayableAmount += row.PayableAmount
		summary.ReceptionCount += row.ReceptionCount
	}

	return &UserPerformanceByReceptionResponse{
		Rows:    rows,
		Summary: summary,
		Meta:    meta,
	}, nil
}

// UserPerformanceByTransaction گزارش کارکرد کاربران بر اساس تراکنش مالی را برمی‌گرداند.
func (s *Service) UserPerformanceByTransaction(q UserPerformanceByTransactionQuery) (*UserPerformanceByTransactionResponse, error) {
	from, to, userIDs, err := parseUserPerformanceBase(q.UserPerformanceBaseQuery)
	if err != nil {
		return nil, err
	}

	infos, err := s.repo.FindUserInfos(userIDs)
	if err != nil {
		return nil, apperror.New("DB_ERROR", "خطا در خواندن اطلاعات کاربران.", err.Error(), 500)
	}
	sortedIDs := sortUserIDsByType(userIDs, infos)

	meta, err := s.buildUserPerformanceMeta("by_transaction", q.FromDate, q.ToDate, q.FromTime, q.ToTime, sortedIDs, q.SeparateByPatient)
	if err != nil {
		return nil, err
	}

	if q.SeparateByPatient {
		sections := make([]UserPerformanceTransactionSection, 0, len(sortedIDs))
		sectionSummaries := make([]UserPerformanceTransactionSummary, 0, len(sortedIDs))
		for _, userID := range sortedIDs {
			detailRows, err := s.repo.ListUserPerformanceTransactions(from, to, userID)
			if err != nil {
				return nil, apperror.New("DB_ERROR", "خطا در محاسبه گزارش.", err.Error(), 500)
			}
			txRows := buildUserTransactionDetailRows(detailRows)
			summary := summarizeUserTransactionDetailRows(txRows)
			sections = append(sections, UserPerformanceTransactionSection{
				UserID:   userID,
				UserName: userDisplayName(userID, infos),
				Rows:     txRows,
				Summary:  summary,
			})
			sectionSummaries = append(sectionSummaries, summary)
		}
		return &UserPerformanceByTransactionResponse{
			Summary:  mergeUserTransactionSummaries(sectionSummaries),
			Sections: sections,
			Meta:     meta,
		}, nil
	}

	rows := make([]UserPerformanceTransactionRow, 0, len(sortedIDs))
	var summary UserPerformanceTransactionSummary
	summary.RowLabel = "جمع کل"
	for _, userID := range sortedIDs {
		agg, err := s.repo.AggregateUserPerformanceByTransaction(from, to, userID)
		if err != nil {
			return nil, apperror.New("DB_ERROR", "خطا در محاسبه گزارش.", err.Error(), 500)
		}
		row := buildUserTransactionRow(userID, userDisplayName(userID, infos), agg)
		rows = append(rows, row)
		summary.TotalReceived += row.TotalReceived
		summary.TotalPaid += row.TotalPaid
		summary.ReceivedTxCount += row.ReceivedTxCount
		summary.PaidTxCount += row.PaidTxCount
		summary.TotalTxCount += row.TotalTxCount
	}

	return &UserPerformanceByTransactionResponse{
		Rows:    rows,
		Summary: summary,
		Meta:    meta,
	}, nil
}

File: frontend/src/modules/reports/definitions/userPerformance.ts
import type { ReportDefinition, ReportMetaItem } from '@/platform/reports';
import { formatCurrency, formatGregorianDate } from '@/platform/reports';
import type {
  UserPerformanceMeta,
  UserPerformanceReceptionDetailRow,
  UserPerformanceReceptionRow,
  UserPerformanceReportType,
  UserPerformanceTransactionDetailRow,
  UserPerformanceTransactionRow,
} from '../types';

/** ستون ردیف با شماره‌گذاری خودکار */
const rowNumberColumn = {
  key: 'row_number' as const,
  title: 'ردیف',
  width: 5,
  format: 'number' as const,
  formatter: (_v: unknown, _r: unknown, index: number) => String(index + 1),
};

/** ReportDefinition — کارکرد بر اساس پذیرش (تجمیع کاربر) */
export const userPerformanceByReceptionDefinition: ReportDefinition<UserPerformanceReceptionRow> = {
  id: 'user-performance-by-reception',
  title: 'گزارش کارکرد کاربران — بر اساس پذیرش',
  page: { size: 'A4', orientation: 'portrait', margin: 12 },
  columns: [
    rowNumberColumn,
    {
      key: 'user_name',
      title: 'نام و نام خانوادگی کاربر',
      width: 18,
      align: 'right',
      format: 'text',
    },
    {
      key: 'service_amount',
      title: 'نرخ خدمات',
      width: 10,
      format: 'currency',
      formatter: (v) => formatCurrency(v),
    },
    {
      key: 'service_tariff',
      title: 'تعرفه خدمات',
      width: 10,
      format: 'currency',
      formatter: (v) => formatCurrency(v),
    },
    {
      key: 'organization_share',
      title: 'سهم سازمان',
      width: 10,
      format: 'currency',
      formatter: (v) => formatCurrency(v),
    },
    {
      key: 'supplementary_share',
      title: 'سهم بیمه تکمیلی',
      width: 11,
      format: 'currency',
      formatter: (v) => formatCurrency(v),
    },
    {
      key: 'subsidy_share',
      title: 'سهم یارانه',
      width: 9,
      format: 'currency',
      formatter: (v) => formatCurrency(v),
    },
    {
      key: 'payable_amount',
      title: 'قابل پرداخت',
      width: 10,
      format: 'currency',
      formatter: (v) => formatCurrency(v),
    },
    {
      key: 'reception_count',
      title: 'تعداد پذیرش',
      width: 9,
      format: 'number',
    },
  ],
  summary: {
    label: 'جمع کل',
    fields: [
      { key: 'service_amount', label: 'جمع نرخ', format: 'currency' },
      { key: 'service_tariff', label: 'جمع تعرفه', format: 'currency' },
      { key: 'organization_share', label: 'جمع سهم سازمان', format: 'currency' },
      { key: 'supplementary_share', label: 'جمع تکمیلی', format: 'currency' },
      { key: 'subsidy_share', label: 'جمع یارانه', format: 'currency' },
      { key: 'payable_amount', label: 'جمع قابل پرداخت', format: 'currency' },
      { key: 'reception_count', label: 'تعداد پذیرش', format: 'number' },
    ],
  },
  header: { showGeneratedAt: true },
  footer: { text: 'صفحات به‌صورت خودکار بر اساس اندازه A4 شکسته می‌شوند.' },
  toolbar: { print: true, pdf: true, excel: true, pageSetup: true },
  export: { sheetName: 'کارکرد پذیرش', includeSections: true },
};

/** ReportDefinition — کارکرد بر اساس پذیرش (تفکیک بیمار) */
export const userPerformanceReceptionDetailDefinition: ReportDefinition<UserPerformanceReceptionDetailRow> =
  {
    id: 'user-performance-reception-detail',
    title: 'گزارش کارکرد کاربران — ریز پذیرش‌ها',
    page: { size: 'A4', orientation: 'portrait', margin: 12 },
    columns: [
      rowNumberColumn,
      { key: 'reception_id', title: 'شماره پذیرش', width: 9, format: 'number' },
      {
        key: 'patient_name',
        title: 'نام و نام خانوادگی بیمار',
        width: 14,
        align: 'right',
        format: 'text',
      },
      { key: 'national_code', title: 'کد ملی بیمار', width: 10, format: 'text' },
      {
        key: 'service_amount',
        title: 'نرخ خدمت',
        width: 9,
        format: 'currency',
        formatter: (v) => formatCurrency(v),
      },
      {
        key: 'service_tariff',
        title: 'تعرفه خدمت',
        width: 9,
        format: 'currency',
        formatter: (v) => formatCurrency(v),
      },
      {
        key: 'organization_share',
        title: 'سهم سازمان',
        width: 9,
        format: 'currency',
        formatter: (v) => formatCurrency(v),
      },
      {
        key: 'supplementary_share',
        title: 'سهم بیمه تکمیلی',
        width: 10,
        format: 'currency',
        formatter: (v) => formatCurrency(v),
      },
      {
        key: 'subsidy_share',
        title: 'سهم یارانه',
        width: 8,
        format: 'currency',
        formatter: (v) => formatCurrency(v),
      },
      {
        key: 'payable_amount',
        title: 'قابل پرداخت',
        width: 9,
        format: 'currency',
        formatter: (v) => formatCurrency(v),
      },
      {
        key: 'services_received',
        title: 'خدمات پذیرش شده',
        width: 20,
        align: 'right',
        format: 'text',
      },
    ],
    summary: {
      label: 'جمع',
      fields: [
        { key: 'service_amount', label: 'جمع نرخ', format: 'currency' },
        { key: 'service_tariff', label: 'جمع تعرفه', format: 'currency' },
        { key: 'organization_share', label: 'جمع سهم سازمان', format: 'currency' },
        { key: 'supplementary_share', label: 'جمع تکمیلی', format: 'currency' },
        { key: 'subsidy_share', label: 'جمع یارانه', format: 'currency' },
        { key: 'payable_amount', label: 'جمع قابل پرداخت', format: 'currency' },
        { key: 'reception_count', label: 'تعداد پذیرش', format: 'number' },
      ],
    },
    header: { showGeneratedAt: true },
    footer: { text: 'صفحات به‌صورت خودکار بر اساس اندازه A4 شکسته می‌شوند.' },
    toolbar: { print: true, pdf: true, excel: true, pageSetup: true },
    export: { sheetName: 'ریز پذیرش', includeSections: true },
  };

/** ReportDefinition — کارکرد بر اساس تراکنش مالی (تجمیع کاربر) */
export const userPerformanceByTransactionDefinition: ReportDefinition<UserPerformanceTransactionRow> = {
  id: 'user-performance-by-transaction',
  title: 'گزارش کارکرد کاربران — بر اساس تراکنش مالی',
  page: { size: 'A4', orientation: 'portrait', margin: 12 },
  columns: [
    rowNumberColumn,
    {
      key: 'user_name',
      title: 'نام و نام خانوادگی کاربر',
      width: 20,
      align: 'right',
      format: 'text',
    },
    {
      key: 'total_received',
      title: 'دریافتی کل',
      width: 12,
      format: 'currency',
      formatter: (v) => formatCurrency(v),
    },
    {
      key: 'total_paid',
      title: 'پرداختی کل',
      width: 12,
      format: 'currency',
      formatter: (v) => formatCurrency(v),
    },
    {
      key: 'received_tx_count',
      title: 'تعداد تراکنش دریافتی',
      width: 12,
      format: 'number',
    },
    {
      key: 'paid_tx_count',
      title: 'تعداد تراکنش پرداختی',
      width: 12,
      format: 'number',
    },
    {
      key: 'total_tx_count',
      title: 'تعداد کل تراکنش‌ها',
      width: 12,
      format: 'number',
    },
  ],
  summary: {
    label: 'جمع کل',
    fields: [
      { key: 'total_received', label: 'جمع دریافتی', format: 'currency' },
      { key: 'total_paid', label: 'جمع پرداختی', format: 'currency' },
      { key: 'received_tx_count', label: 'تراکنش دریافتی', format: 'number' },
      { key: 'paid_tx_count', label: 'تراکنش پرداختی', format: 'number' },
      { key: 'total_tx_count', label: 'کل تراکنش‌ها', format: 'number' },
    ],
  },
  header: { showGeneratedAt: true },
  footer: { text: 'صفحات به‌صورت خودکار بر اساس اندازه A4 شکسته می‌شوند.' },
  toolbar: { print: true, pdf: true, excel: true, pageSetup: true },
  export: { sheetName: 'کارکرد تراکنش', includeSections: true },
};

/** ReportDefinition — کارکرد بر اساس تراکنش (ریز تراکنش‌ها) */
export const userPerformanceTransactionDetailDefinition: ReportDefinition<UserPerformanceTransactionDetailRow> =
  {
    id: 'user-performance-transaction-detail',
    title: 'گزارش کارکرد کاربران — ریز تراکنش‌ها',
    page: { size: 'A4', orientation: 'portrait', margin: 12 },
    columns: [
      rowNumberColumn,
      { key: 'transaction_id', title: 'شماره تراکنش', width: 10, format: 'number' },
      { key: 'file_number', title: 'شماره پرونده', width: 10, format: 'text' },
      {
        key: 'patient_name',
        title: 'نام و نام خانوادگی بیمار',
        width: 16,
        align: 'right',
        format: 'text',
      },
      { key: 'national_code', title: 'کد ملی بیمار', width: 10, format: 'text' },
      {
        key: 'received',
        title: 'دریافتی',
        width: 12,
        format: 'currency',
        formatter: (v) => formatCurrency(v),
      },
      {
        key: 'paid',
        title: 'پرداختی',
        width: 12,
        format: 'currency',
        formatter: (v) => formatCurrency(v),
      },
    ],
    summary: {
      label: 'جمع',
      fields: [
        { key: 'total_received', label: 'جمع دریافتی', format: 'currency' },
        { key: 'total_paid', label: 'جمع پرداختی', format: 'currency' },
        { key: 'received_tx_count', label: 'تراکنش دریافتی', format: 'number' },
        { key: 'paid_tx_count', label: 'تراکنش پرداختی', format: 'number' },
        { key: 'total_tx_count', label: 'کل تراکنش‌ها', format: 'number' },
      ],
    },
    header: { showGeneratedAt: true },
    footer: { text: 'صفحات به‌صورت خودکار بر اساس اندازه A4 شکسته می‌شوند.' },
    toolbar: { print: true, pdf: true, excel: true, pageSetup: true },
    export: { sheetName: 'ریز تراکنش', includeSections: true },
  };

/** meta rows برای گزارش کارکرد کاربران */
export function buildUserPerformanceMetaRows(meta: UserPerformanceMeta): ReportMetaItem[][] {
  const rows: ReportMetaItem[][] = [
    [
      { label: 'از تاریخ', value: `${formatGregorianDate(meta.from_date)} — ${meta.from_time}` },
      { label: 'تا تاریخ', value: `${formatGregorianDate(meta.to_date)} — ${meta.to_time}` },
    ],
    [
      {
        label: 'کاربران',
        value: meta.user_names.length ? meta.user_names.join('، ') : '—',
      },
    ],
  ];

  if (meta.separate_by_patient) {
    rows.push([{ label: 'نمایش', value: 'لیست به تفکیک بیماران / پرونده' }]);
  }

  return rows;
}

/** definition مناسب بر اساس نوع گزارش */
export function getUserPerformanceDefinition(
  type: UserPerformanceReportType,
  separateByPatient: boolean,
) {
  if (type === 'by_reception') {
    return separateByPatient
      ? userPerformanceReceptionDetailDefinition
      : userPerformanceByReceptionDefinition;
  }
  return separateByPatient
    ? userPerformanceTransactionDetailDefinition
    : userPerformanceByTransactionDefinition;
}

/** sections را برای ReportViewer آماده می‌کند */
export function mapUserSections<
  T extends { user_id: number; user_name: string; rows: object[]; summary: object },
>(sections: T[] | undefined) {
  return sections?.map((s) => ({
    id: String(s.user_id),
    title: s.user_name,
    rows: s.rows,
    summary: s.summary as Record<string, unknown>,
  }));
}
